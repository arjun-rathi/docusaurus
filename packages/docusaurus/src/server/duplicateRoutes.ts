/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {ReportingSeverity, RouteConfig} from '@docusaurus/types';
import {reportMessage} from '@docusaurus/utils';
import {getAllFinalRoutes} from './utils';

function getAllDuplicateRoutes(pluginsRouteConfigs: RouteConfig[]): string[] {
  const allRoutes: string[] = getAllFinalRoutes(pluginsRouteConfigs).map(
    (routeConfig) => routeConfig.path,
  );
  const seenRoutes = new Set<string>();
  return allRoutes.filter((route) => {
    if (seenRoutes.has(route)) {
      return true;
    }
    seenRoutes.add(route);
    return false;
  });
}

function getDuplicateRoutesMessage(allDuplicateRoutes: string[]): string {
  const message = allDuplicateRoutes
    .map(
      (duplicateRoute) =>
        `- Attempting to create page at ${duplicateRoute}, but a page already exists at this route.`,
    )
    .join('\n');
  return message;
}

export function handleDuplicateRoutes(
  pluginsRouteConfigs: RouteConfig[],
  onDuplicateRoutes: ReportingSeverity,
): void {
  if (onDuplicateRoutes === 'ignore') {
    return;
  }
  const duplicatePaths: string[] = getAllDuplicateRoutes(pluginsRouteConfigs);
  const message: string = getDuplicateRoutesMessage(duplicatePaths);
  if (message) {
    const finalMessage = `Duplicate routes found!
${message}
This could lead to non-deterministic routing behavior.`;
    reportMessage(finalMessage, onDuplicateRoutes);
  }
}
