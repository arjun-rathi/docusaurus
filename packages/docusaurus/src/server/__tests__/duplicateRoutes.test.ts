/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {jest} from '@jest/globals';
import {handleDuplicateRoutes} from '../duplicateRoutes';
import type {RouteConfig} from '@docusaurus/types';

const routes: RouteConfig[] = [
  {
    path: '/',
    component: '',
    routes: [
      {path: '/search', component: ''},
      {path: '/sameDoc', component: ''},
    ],
  },
  {
    path: '/',
    component: '',
    routes: [
      {path: '/search', component: ''},
      {path: '/sameDoc', component: ''},
      {path: '/uniqueDoc', component: ''},
    ],
  },
  {
    path: '/',
    component: '',
  },
  {
    path: '/',
    component: '',
  },
  {
    path: '/',
    component: '',
  },
];

describe('handleDuplicateRoutes', () => {
  it('works', () => {
    expect(() => {
      handleDuplicateRoutes(routes, 'throw');
    }).toThrowErrorMatchingSnapshot();
    const consoleMock = jest.spyOn(console, 'log').mockImplementation(() => {});
    handleDuplicateRoutes(routes, 'ignore');
    expect(consoleMock).toBeCalledTimes(0);
  });
});
