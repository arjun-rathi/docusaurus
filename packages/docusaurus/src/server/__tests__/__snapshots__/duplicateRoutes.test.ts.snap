// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleDuplicateRoutes works 1`] = `
"Duplicate routes found!
- Attempting to create page at /search, but a page already exists at this route.
- Attempting to create page at /sameDoc, but a page already exists at this route.
- Attempting to create page at /, but a page already exists at this route.
- Attempting to create page at /, but a page already exists at this route.
This could lead to non-deterministic routing behavior."
`;
